### PR TITLE
Update certificate validation logic (EXPOSUREAPP-10447)

### PR DESCRIPTION
**Pull request by jurajkusnier**
- updated validation
- error dialog
## Testing
The following presets can be used for testing the feature against the mock server:

- should be blocked:
    - `cwa vc gen --preset 60`
    - `cwa vc gen --preset 61`
    - `cwa vc gen --preset 63`

- should not be blocked:
    - `cwa vc gen --preset 62`

## Jira Link
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-10238

### CODE REVIEW

**Review by CV113 — APPROVED** [review]
UA approved

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/ui/onboarding/CovidCertificateOnboardingViewModel.kt [hunk 1]
```diff
@@ -26,7 +29,12 @@ class CovidCertificateOnboardingViewModel @AssistedInject constructor(
         covidCertificateSettings.isOnboarded.update { true }
         val event = if (dccQrCode != null) {
             try {
-                val containerId = dccQrCodeHandler.handleQrCode(dccQrCode)
+                val containerId = dccQrCodeHandler.handleQrCode(
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/ui/onboarding/CovidCertificateOnboardingViewModel.kt:32]: what about injecting the AppConfig in `DccQrCodeHandler` itself, so you don't need to inject the AppConfigProvider everywhere you need to call the Handler

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/common/certificate/CwaCovidCertificate.kt [hunk 2]
```diff
@@ -105,6 +107,17 @@ interface CwaCovidCertificate : Recyclable {
                 .registerSubtype(ExpiringSoon::class.java, "ExpiringSoon")
                 .registerSubtype(Expired::class.java, "Expired")
                 .registerSubtype(Invalid::class.java, "Invalid")
+                .registerSubtype(Blocked::class.java, "Blocked")
+        }
+
+        override fun equals(other: Any?): Boolean {
+            if (this is Blocked && other is Blocked) return true
+            if (this is Recycled && other is Recycled) return true
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/common/certificate/CwaCovidCertificate.kt:115]: why do you need special handling for these cases ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/ui/onboarding/CovidCertificateOnboardingViewModel.kt (hunk 1 above)
Comment (mtwalli) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/ui/onboarding/CovidCertificateOnboardingViewModel.kt:32]: I would rather do that in the `BlocklistValidator` itself

**Comment by mtwalli** [comment]
All certificates in the description are `VC` do you have some `TC, RC` that can be blocked?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/qrcode/ui/ScannerExtensions.kt [hunk 3]
```diff
@@ -37,6 +37,12 @@ fun Throwable.toQrCodeErrorDialogBuilder(context: Context): MaterialAlertDialogB
                         context.openUrl(R.string.error_dcc_already_registered_faq_link)
                     }
                 }
+                throwable.isBlocked -> {
+                    setTitle(R.string.error_dcc_in_blocklist_title)
+                    setNegativeButton(R.string.error_dcc_in_blocklist_faq) { _, _ ->
+                        context.openUrl(R.string.dcc_signature_validation_dialog_faq_link)
+                    }
+                }
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/qrcode/ui/ScannerExtensions.kt:45]: Dialog buttons should be reversed in order
Screenshot 2021-11-09 at 15 37 14

Look at figma https://www.figma.com/file/KuDEcMSREg3VhZRhC3FXTY/COVID19_Master_2.11?node-id=37357%3A27922

Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/qrcode/ui/ScannerExtensions.kt:45]: AFIK that's iOS dialog, and we should rather stick to our material design guidelines where the button order is as follows: 
image

Comment (mtwalli) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/qrcode/ui/ScannerExtensions.kt:45]: OK, btw I think there is an issue here from before , we should use `setNeutralButton` as other dialogs ,but apparently it is overridden by  ` .setNeutralButton(null, null) // Remove details` and this removes any link in the dialogs. can you please fix it? I think the following would fix it 
```kotlin
    return toErrorDialogBuilder(context).apply {
        if (throwable is InvalidHealthCertificateException) {
            when {
                throwable.isCertificateInvalid ->
                    setNeutralButton(R.string.error_button_dcc_faq) { _, _ ->
                        context.openUrl(R.string.error_button_dcc_faq_link)
                    }

                throwable.isSignatureInvalid -> {
                    setTitle(R.string.dcc_signature_validation_dialog_title)
                    setNeutralButton(R.string.dcc_signature_validation_dialog_faq_button) { _, _ ->
                        context.openUrl(R.string.dcc_signature_validation_dialog_faq_link)
                    }
                }
                throwable.isAlreadyRegistered -> {
                    setNeutralButton(R.string.error_button_dcc_faq) { _, _ ->
                        context.openUrl(R.string.error_dcc_already_registered_faq_link)
                    }
                }
                throwable.isBlocked -> {
                    setTitle(R.string.error_dcc_in_blocklist_title)
                    setNeutralButton(R.string.error_dcc_in_blocklist_faq) { _, _ ->
                        context.openUrl(R.string.dcc_signature_validation_dialog_faq_link)
                    }
                }
            }
        } else {
            setNeutralButton(null, null) // Remove details
        }
    }
```

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/common/certificate/CwaCovidCertificate.kt (hunk 2 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/common/certificate/CwaCovidCertificate.kt:115]: comparing state with deserialized state didn't work as expected in https://github.com/corona-warn-app/cwa-app-android/blob/747cf79192f218129e4d9fbdad8de8fb8d8ab9ca/Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/vaccination/core/repository/storage/VaccinationContainer.kt#L175

even if `state` and `lastSeenStateChange` were objects of the same type `Blocked`, equals returned false. It works for data classes.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/qrcode/ui/ScannerExtensions.kt (hunk 3 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/qrcode/ui/ScannerExtensions.kt:45]: other `InvalidHealthCertificateException` shouldn't have an FAQ button according to the Figma design. It has been changed by this PR https://github.com/corona-warn-app/cwa-app-android/pull/4171 by adding `setNeutralButton(null, null)` but I can probably remove `setNeutralButton()` directly from `toErrorDialogBuilder()` method when we don't need it anymore

Comment (mtwalli) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/qrcode/ui/ScannerExtensions.kt:45]: > other InvalidHealthCertificateException shouldn't have an FAQ button according to the Figma design.
Where do you see that ?
#4171  was only to remove the `details` button not FAQ button ,I think this wasn't intended ,right?
/cc @BMItr

Comment (BMItr) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/qrcode/ui/ScannerExtensions.kt:45]: yes, #4171 was only to remove the details button (and stay on the screen to scan again)

Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/qrcode/ui/ScannerExtensions.kt]: updated

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/ui/onboarding/CovidCertificateOnboardingViewModel.kt (hunk 1 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/ui/onboarding/CovidCertificateOnboardingViewModel.kt]: it's not used there anymore

**Pull request merged**